File: PRACTICAS/CAMBIAR_RUBY/COMPROBACIONES JAVA/src/civitas_p1/Tablero.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas_p1;
import java.util.*;

/**
 *
 * @author angelsolano
 */

public class Tablero {
    
    private int numCasillaCarcel; 
    private ArrayList<Casilla> casillas;
    private int porSalida;
    private boolean tieneJuez; 
    
    private boolean correcto(){
        return casillas.size() > numCasillaCarcel && tieneJuez;
    }
    
    private boolean correcto(int numCasilla){
        return correcto() && numCasilla >= 0 && numCasilla < casillas.size();
    }
    
    public Tablero(int numCasillaCarcel){
        
        if (numCasillaCarcel > 1)
            this.numCasillaCarcel = numCasillaCarcel;
        else
            this.numCasillaCarcel = 1;
        
        Casilla salida;
        salida = new Casilla("Salida");

        casillas = new ArrayList<Casilla>();
        casillas.add(salida);
        porSalida = 0;
        tieneJuez = false;
        
    }
    
    int getCarcel(){
        return numCasillaCarcel;
    }
    
    int getPorSalida(){
        
        if (porSalida > 0){
            porSalida --;
            return porSalida + 1;
        }
        else
            return porSalida;
     
    }
    
    void añadeCasilla(Casilla casilla){
        
        Casilla carcel;
        carcel = new Casilla("Carcel");
        
        if (casillas.size() == numCasillaCarcel){
            casillas.add(carcel);
            casillas.add(casilla);           
        }
        else{
            
            casillas.add(casilla);
            
            if (casillas.size() == numCasillaCarcel)
                casillas.add(carcel);
            
        }
            
    }
    
    void añadeJuez(){
        
        if (!tieneJuez){
            
            Casilla juez;
            juez = new Casilla(numCasillaCarcel,"Juez");
            
            
            añadeCasilla(juez);
            tieneJuez = true;
            
        }
    }
    
    Casilla getCasilla(int numCasilla){
        
        if (correcto(numCasilla))
            return casillas.get(numCasilla);
        else
            return null;
        
    }
    
    int nuevaPosicion(int actual, int tirada){
        
        if(!correcto())
            return -1;
        else{
            
            int nueva_pos;
            
            nueva_pos = (actual + tirada) % casillas.size();
            
            if (nueva_pos != (actual + tirada))
                porSalida ++;
            
            return nueva_pos;
            
        }
                
    }
    
    int calcularTirada(int origen, int destino){
        
        int tirada = destino - origen;
        
        if (tirada < 0)
            return tirada + casillas.size();
        else
            return tirada;
        
        
    }
    
    public ArrayList<Casilla> getCasillas(){
        return casillas;
    }
    
    
    
}

File: PRACTICAS/CAMBIAR_RUBY/RUBY2/Civitas/pkg/RubyApplication1-0.0.1/lib/Tablero.rb
module Civitas
  class Tablero
    attr_accessor :numCasillaCarcel, :casillas, :porSalida, :tieneJuez
    def initialize (carcel, casillas, salida, juez)
      @numCasillaCarcel = carcel
      @casillas = Array.new
      for i in 0 .. casillas.length
        @casillas.push(casillas[i])
      end
      @porSalida = salida
      @tieneJuez = juez
    end
    
    def correcto
      correct = true
      if(@casillas.length <= @numCasillaCarcel)
        correct = false
      end
      if (@tieneJuez == false)
        correct = false
      end
      return correct
    end
    
    def getCarcel
      return @numCasillaCarcel
    end
    
    def getPorSalida
      if (@porSalida > 0)
        @porSalida = @porSalida -1
        return @porSalida +1
      else
        return @porSalida
      end
    end
    
    def aniadeCasilla (casilla)
      if(@casillas == nil)
        @casillas = Array.new
      end
      @casillas.push(casilla)
    end
    
    def aniadeJuez
      if (@tieneJuez == false)
        @juez.setNombre("Juez")
        @casillas.push(@juez)
        @tieneJuez = true
      end
    end
    
    def getCasilla (numCasilla)
      return @casillas[numCasilla]
    end
    
    def nuevaPosicion (actual, tirada)
      if (correcto == false)
        return -1;
      else
        @pos = actual + tirada
        if (@pos > @casillas.length)
          @pos = @pos - @casillas.length
        end
      end
      return @pos
    end
    
    def calcularTirada (origen, destino)
      @tirada = destino - origen
      if (tirada < 0)
        @tirada = @tirada + @casillas.length
      end
      return @tirada
    end
    
    def borrar_casilla (ip)
      @casillas.delete_at(ip)
    end
    
    private :correcto
  end
end
